pkgs/jni/lib/src/jarray.dart

```diff
@@ -193,122 +384,546 @@ extension BoolArray on JArray<jboolean> {
   }
 
   void setRange(int start, int end, Iterable<int> iterable,
       [int skipCount = 0]) {
     RangeError.checkValidRange(start, end, length);
     final rangeLength = end - start;
-    _allocate<JShortMarker>(sizeOf<JShortMarker>() * rangeLength, (ptr) {
+    _allocate<JByteMarker>(sizeOf<JByteMarker>() * rangeLength, (ptr) {
       ptr
           .asTypedList(rangeLength)
           .setRange(0, rangeLength, iterable, skipCount);
-      Jni.env.SetShortArrayRegion(reference.pointer, start, rangeLength, ptr);
+      Jni.env.SetByteArrayRegion(reference.pointer, start, rangeLength, ptr);
     });
   }
+
+  @override
+  Iterator<int> get iterator => _JArrayIterator(this);
 }
 
-extension IntArray on JArray<jint> {
-  int operator [](int index) {
-    RangeError.checkValidIndex(index, this);
-    return Jni.env.GetIntArrayElement(reference.pointer, index);
-  }
+final class JCharArrayNullableType extends JObjType<JCharArray?> {
+  @internal
+  const JCharArrayNullableType();
 
-  void operator []=(int index, int value) {
-    RangeError.checkValidIndex(index, this);
-    Jni.env.SetIntArrayElement(reference.pointer, index, value);
-  }
+  @internal
+  @override
+  String get signature => '[C';
 
-  Int32List getRange(int start, int end, {Allocator allocator = malloc}) {
-    RangeError.checkValidRange(start, end, length);
-    final rangeLength = end - start;
+  @internal
+  @override
+  JCharArray? fromReference(JReference reference) =>
+      reference.isNull ? null : JCharArray.fromReference(reference);
+
+  @internal
+  @override
+  JObjType get superType => const JObjectNullableType();
+
+  @internal
+  @override
+  JObjType<JCharArray?> get nullableType => this;
+
+  @internal
+  @override
+  final int superCount = 1;
+
+  @override
+  int get hashCode => (JCharArrayNullableType).hashCode;
+
+  @override
+  bool operator ==(Object other) {
+    return other.runtimeType == JCharArrayNullableType &&
+        other is JCharArrayNullableType;
+  }
+}
+
+final class JCharArrayType extends JObjType<JCharArray> {
+  @internal
+  const JCharArrayType();
+
+  @internal
+  @override
+  String get signature => '[C';
+
+  @internal
+  @override
+  JCharArray fromReference(JReference reference) =>
+      JCharArray.fromReference(reference);
+
+  @internal
+  @override
+  JObjType get superType => const JObjectType();
+
+  @internal
+  @override
+  JObjType<JCharArray?> get nullableType => const JCharArrayNullableType();
+
+  @internal
+  @override
+  final int superCount = 1;
+
+  @override
+  int get hashCode => (JCharArrayType).hashCode;
+
+  @override
+  bool operator ==(Object other) {
+    return other.runtimeType == JCharArrayType && other is JCharArrayType;
+  }
+}
+
+/// `JCharArray` is a 16-bit integer array.
+///
+/// Due to variable length encoding, the  number of code units is not equal to
+/// the number of characters.
+class JCharArray extends JObject with Iterable<int> {
```

Comment:
This is a lot of duplicated code with `JShortArray`, `JIntArray` and `JLongArray`. They Dart API should be identical, can some of the API and the implementation be shared?

If not, consider generating this code with a small generator so that any edits in the future don't have to be done in multiple places. (See runtime/tools/ffi/sdk_lib_ffi_generator.dart in the SDK, which might be of use.)

Reply:
Will do in the future.